Reading/Writing GeoJSON via GEOSCoordSequence_copy(From|To)Buffer (#17)

sanak opened the pull request:
Hi @chrispahm,
I implemented #17 on this PR branch, and I confirmed that `npm run performance` result was a bit (1.5x) faster than `wkb(wkx)` one, so just for your information. 😄

| Test Pattern                      | Current wkb(wkx) | Use this PR | @turf/turf (2 times) |
|--------------------------|-------------------|-------------|--------------------|
| Testing feature Simple     | 29 ms |  **18 ms** |  54/56 ms |
| Testing feature Australia  | 5204 ms | **3098 ms** | 6473/6465 ms |
| Real life performance test | 300 ms | **188 ms** | 244/238 ms |
| Buffering each geometry individually | 258 ms | **134 ms** | 141/144 ms |

`Use this PR` needs the following changes in `test/performance.mjs`.
```diff
--- a/test/performance.mjs
+++ b/test/performance.mjs
@@ -2,7 +2,7 @@ import { performance } from 'perf_hooks'
 import * as turf from '@turf/turf'
 import { readFileSync } from 'fs'
 import initGeosJs from '../src/index.mjs'
-import geosBuffer from '../docs/assets/buffer.js'
+import geosBuffer from '../docs/assets/buffer2.js'
```

<details>

### Current wkb(wkx)

```sh
% npm run performance

> geos-wasm@1.1.4 performance
> node test/performance.mjs

Testing feature Simple
GEOS buffer method took 29 milliseconds to run 100 times.
Average time per iteration: 0 milliseconds.
@turf/turf buffer method took 54 milliseconds to run 100 times.
Average time per iteration: 1 milliseconds.
GEOS buffer method was 1.86x faster than @turf/turf buffer method.

Testing feature Australia
GEOS buffer method took 5204 milliseconds to run 100 times.
Average time per iteration: 52 milliseconds.
@turf/turf buffer method took 6473 milliseconds to run 100 times.
Average time per iteration: 65 milliseconds.
GEOS buffer method was 1.24x faster than @turf/turf buffer method.

Real life performance test
Buffering all geometries at once
GEOS buffer method took 300 milliseconds to run 1 times.
Average time per iteration: 300 milliseconds.
@turf/turf buffer method took 244 milliseconds to run 1 times.
Average time per iteration: 244 milliseconds.
@turf/turf buffer method was 1.23x faster than GEOS buffer method.

Buffering each geometry individually
GEOS buffer method took 258 milliseconds to run 1 times.
Average time per iteration: 258 milliseconds.
@turf/turf buffer method took 141 milliseconds to run 1 times.
Average time per iteration: 141 milliseconds.
@turf/turf buffer method was 1.83x faster than GEOS buffer method.
```

### Use this PR version

```sh
% npm run performance

> geos-wasm@1.1.4 performance
> node test/performance.mjs

Testing feature Simple
GEOS buffer method took 18 milliseconds to run 100 times.
Average time per iteration: 0 milliseconds.
@turf/turf buffer method took 56 milliseconds to run 100 times.
Average time per iteration: 1 milliseconds.
GEOS buffer method was 3.11x faster than @turf/turf buffer method.

Testing feature Australia
GEOS buffer method took 3098 milliseconds to run 100 times.
Average time per iteration: 31 milliseconds.
@turf/turf buffer method took 6465 milliseconds to run 100 times.
Average time per iteration: 65 milliseconds.
GEOS buffer method was 2.09x faster than @turf/turf buffer method.

Real life performance test
Buffering all geometries at once
GEOS buffer method took 188 milliseconds to run 1 times.
Average time per iteration: 188 milliseconds.
@turf/turf buffer method took 238 milliseconds to run 1 times.
Average time per iteration: 238 milliseconds.
GEOS buffer method was 1.27x faster than @turf/turf buffer method.

Buffering each geometry individually
GEOS buffer method took 134 milliseconds to run 1 times.
Average time per iteration: 134 milliseconds.
@turf/turf buffer method took 144 milliseconds to run 1 times.
Average time per iteration: 144 milliseconds.
GEOS buffer method was 1.07x faster than @turf/turf buffer method.
```

</details>

Before the implementation, I wrote some Coordinate Sequence and Geometry Constructor tests to be familiar with the usage and I fixed/adjusted some parts, but those can be separated to another PR, so if you feel that this changes are too much, let me know about that.

chrispahm commented on docs/assets/buffer2.js:
This is awesome! I'd only suggest to rename this file to something like buffer_GEOSCoordSequence, and the other file to buffer_wkb, so it's clearer what the difference between the two is!

sanak replied:
Oh, yes. `buffer2.js` was just temporary name. 😄
If the original file rename to `buffer_wkb.js`, how about similar small capital short name `buffer_coordseq.js` ?

chrispahm replied:
Yes, sounds good!

chrispahm commented on docs/assets/buffer2.js:
```diff
@@ -0,0 +1,332 @@
+  // const geomPtr = GEOSGeomFromWKT(stringify(projected));
+  // const wkb = Geometry.parseGeoJSON(projected).toWkb()
+  // const geomPtr = GEOSGeomFromWKB(wkb)
+  const geomPtr = jsonGeomToGeosGeom(projected)
```
This is nitpicking, but I'd prefer the function to be called geojsonToGeosGeom, as to me jsonGeom sounds like it's some custom json geometry format

sanak replied:
Okay, sure.
In `src/helpers/geojson.mjs`, there are 2 private (non exported) methods at the bottom, so it may be better to rename as follows.
- `jsonCoordsToGeosCoordSeq` => `geojsonCoordsToGeosCoordSeq`
- `geosCoordSeqToJsonCoords` => `geosCoordSeqToGeojsonCoords`

chrispahm commented on docs/assets/buffer2.js:
```diff
@@ -0,0 +1,332 @@
+  // update the original GeoJSON with the new geometry
+  // const bufferedWkb = GEOSGeomToWKB(bufferPtr)
+  // const buffered = Geometry.parse(bufferedWkb).toGeoJSON()
+  const buffered = geosGeomToJsonGeom(bufferPtr)
```
Same as above, would prefer geosGeomToGeojson

sanak replied:
Same as above my comment.

Thanks a lot!

chrispahm reviewed (changes requested):
Thanks again for this awesome PR! There are a few tiny changes I'd ask for, as outlined in the review comments. Other things that would be nice would be to
- incorporate this method into the docs examples
- Update the FAQ https://chrispahm.github.io/geos-wasm/#what-is-the-preferred-way-to-pass-geometries-between-javascript-and-geos-wasm to reflect this more performant option
- ?

Let me know if you want me to take over any of those tasks!

sanak commented:
@chrispahm Thanks for confirmation!

> Thanks again for this awesome PR! There are a few tiny changes I'd ask for, as outlined in the review comments. Other things that would be nice would be to
> 
> * incorporate this method into the docs examples
> * Update the FAQ https://chrispahm.github.io/geos-wasm/#what-is-the-preferred-way-to-pass-geometries-between-javascript-and-geos-wasm to reflect this more performant option
> * ?
> 
> Let me know if you want me to take over any of those tasks!

Thanks!
Because I am still not familiar with docs/examples and build part, taking over above tasks is really helpful for me. 🙇🙏
I will reply the code comments, so please wait for a while.

sanak commented:
@chrispahm Well, I will take one week summer holidays from tomorrow, so I am okay to do remaining things, if you haven't started update yet.
Though my English writing and understanding docs/examples and build part may take a certain time. 😊

chrispahm commented:
@sanak yes I haven't started working on the docs, so please go ahead! But no pressure, you can also just enjoy your summer holidays 🌞🏖️

sanak commented:
@chrispahm Okay, thanks for confirmation!
And yes, I will enjoy also other things during summer holidays. 😅

sanak commented:
@chrispahm
I am still on about halfway, but I get stuck...

I renamed all files/methods as you mentioned, so those part should be solved, but when I switched from `buffer_wkb.js` to `buffer_coordseq.js` in `docs/components/buffer.vue`, I got the following browser console error from `npm run docs:dev`.
```
geojson.mjs:322 Uncaught TypeError: geos.Module._malloc is not a function
    at geojsonCoordsToGeosCoordSeq (geojson.mjs:322:32)
    at geojsonToGeosGeom (geojson.mjs:65:26)
    at bufferFeature (buffer_coordseq.js:204:19)
    at buffer_coordseq.js:107:31
    at featureEach (index.js:395:11)
    at buffer (buffer_coordseq.js:106:7)
    at updateBuffer (buffer.vue:256:26)
    at Ba.Map.<anonymous> (buffer.vue:296:5)
    at Ba.Map.fire (evented.ts:130:26)
    at Ba.Map._render (map.ts:3194:18)
```

In original `buffer_wkb.js`, `geos` object passed to `GEOSFunctions` via `buffer` method's option property (`GEOS`), but I don't know the best way to pass the object to new helper...

Probably, the easiest way will be adding `geos` object argument at each helper methods (the following diff image),
```diff
--- a/src/helpers/geojson.mjs
+++ b/src/helpers/geojson.mjs
@@ -1,6 +1,4 @@
-import { geos } from '../allCFunctions.mjs'
-
-export function geojsonToGeosGeom (geojson) {
+export function geojsonToGeosGeom (geos, geojson) {
   // assume only 2d (x, y) geometries
   switch (geojson.type) {
     case 'Feature':
@@ -165,7 +163,7 @@ export function geojsonToGeosGeom (geojson) {
   }
 }
 
-export function geosGeomToGeojson (geomPtr) {
+export function geosGeomToGeojson (geos, geomPtr) {
   if (!geomPtr) {
     return null
   }
@@ -312,7 +310,7 @@ export function geosGeomToGeojson (geomPtr) {
   }
 }
 
-function geojsonCoordsToGeosCoordSeq (coords) {
+function geojsonCoordsToGeosCoordSeq (geos, coords) {
   if (!coords || coords.length === 0) {
     return null
   }
@@ -326,7 +324,7 @@ function geojsonCoordsToGeosCoordSeq (coords) {
   return seqPtr
 }
 
-function geosCoordSeqToGeojsonCoords (seqPtr) {
+function geosCoordSeqToGeojsonCoords (geos, seqPtr) {
   if (!seqPtr) {
     return null
   }
```
then pass `geos` argument from client app side, but how do you think about this ?

sanak commented:
@chrispahm About the above `buffer_coordseq.js` issue in `docs/components/buffer.vue`, other solution will be exporting 2 methods via `geos.helpers` object (or namespace).
I prefer this way than above adding `geos` argument` way, but let me know your preference.
```diff
--- a/src/index.mjs
+++ b/src/index.mjs
@@ -1,4 +1,5 @@
 import { initCFunctions, geos } from './allCFunctions.mjs'
+import { geojsonToGeosGeom, geosGeomToGeojson } from './helpers/geojson.mjs'
 import CModule from '../build/package/geos.js'
 
 let geosJsPromise
@@ -30,6 +31,10 @@ export default function initGeosJs (config = {}) {
     }
 
     CModule(geos.Module).then((res) => {
+      geos.helpers = {
+        geojsonToGeosGeom,
+        geosGeomToGeojson
+      }
       resolve(geos)
     })
   })
```
```diff
--- a/docs/assets/buffer_coordseq.js
+++ b/docs/assets/buffer_coordseq.js
@@ -33,7 +33,6 @@ import {
   lengthToRadians,
   earthRadius
 } from '@turf/helpers'
-import { geojsonToGeosGeom, geosGeomToGeojson } from '../../src/helpers/geojson.mjs'
 let GEOSFunctions
 
 /**
@@ -201,7 +200,7 @@ function bufferFeature (geojson, radius, units, steps, endCapStyle, joinStyle, m
     }
   }
   // create a GEOS object from the GeoJSON
-  const geomPtr = geojsonToGeosGeom(projected)
+  const geomPtr = GEOSFunctions.helpers.geojsonToGeosGeom(projected)
   const distance = radiansToLength(lengthToRadians(radius, units), 'meters')
   let bufferPtr
   if (isBufferWithParams) {
@@ -214,7 +213,7 @@ function bufferFeature (geojson, radius, units, steps, endCapStyle, joinStyle, m
     GEOSFunctions.GEOSBufferParams_destroy(bufferParamsPtr)
   }
   // update the original GeoJSON with the new geometry
-  const buffered = geosGeomToGeojson(bufferPtr)
+  const buffered = GEOSFunctions.helpers.geosGeomToGeojson(bufferPtr)
   // destroy the GEOS objects
   GEOSFunctions.GEOSGeom_destroy(geomPtr)
   GEOSFunctions.GEOSGeom_destroy(bufferPtr)
```

chrispahm commented:
> @chrispahm I am still on about halfway, but I get stuck...
> 
> I renamed all files/methods as you mentioned, so those part should be solved, but when I switched from `buffer_wkb.js` to `buffer_coordseq.js` in `docs/components/buffer.vue`, I got the following browser console error from `npm run docs:dev`.
> 
> ```
> geojson.mjs:322 Uncaught TypeError: geos.Module._malloc is not a function
>     at geojsonCoordsToGeosCoordSeq (geojson.mjs:322:32)
>     at geojsonToGeosGeom (geojson.mjs:65:26)
>     at bufferFeature (buffer_coordseq.js:204:19)
>     at buffer_coordseq.js:107:31
>     at featureEach (index.js:395:11)
>     at buffer (buffer_coordseq.js:106:7)
>     at updateBuffer (buffer.vue:256:26)
>     at Ba.Map.<anonymous> (buffer.vue:296:5)
>     at Ba.Map.fire (evented.ts:130:26)
>     at Ba.Map._render (map.ts:3194:18)
> ```
> 
> In original `buffer_wkb.js`, `geos` object passed to `GEOSFunctions` via `buffer` method's option property (`GEOS`), but I don't know the best way to pass the object to new helper...
> 
> Probably, the easiest way will be adding `geos` object argument at each helper methods (the following diff image),
> 
> ```diff
> --- a/src/helpers/geojson.mjs
> +++ b/src/helpers/geojson.mjs
> @@ -1,6 +1,4 @@
> -import { geos } from '../allCFunctions.mjs'
> -
> -export function geojsonToGeosGeom (geojson) {
> +export function geojsonToGeosGeom (geos, geojson) {
>    // assume only 2d (x, y) geometries
>    switch (geojson.type) {
>      case 'Feature':
> @@ -165,7 +163,7 @@ export function geojsonToGeosGeom (geojson) {
>    }
>  }
>  
> -export function geosGeomToGeojson (geomPtr) {
> +export function geosGeomToGeojson (geos, geomPtr) {
>    if (!geomPtr) {
>      return null
>    }
> @@ -312,7 +310,7 @@ export function geosGeomToGeojson (geomPtr) {
>    }
>  }
>  
> -function geojsonCoordsToGeosCoordSeq (coords) {
> +function geojsonCoordsToGeosCoordSeq (geos, coords) {
>    if (!coords || coords.length === 0) {
>      return null
>    }
> @@ -326,7 +324,7 @@ function geojsonCoordsToGeosCoordSeq (coords) {
>    return seqPtr
>  }
>  
> -function geosCoordSeqToGeojsonCoords (seqPtr) {
> +function geosCoordSeqToGeojsonCoords (geos, seqPtr) {
>    if (!seqPtr) {
>      return null
>    }
> ```
> 
> then pass `geos` argument from client app side, but how do you think about this ?

I think passing the `geos` instance in the helper functions is the safest option! For instance, this is how it's done in the [geotoolbox](https://github.com/neocarto/geotoolbox/blob/e2cfa72f4a7ff2604397aa8d9dd5f815fb2443c6/src/gis/buffer.js#L48C33-L48C33). Especially if there are multiple instances of `geos` this should be preferable so the correct instance is chosen!

chrispahm commented:
Let me know when I should review the changes again, then we can merge this PR 😊

sanak commented:
@chrispahm Thanks for the advice! 🙇
> I think passing the geos instance in the helper functions is the safest option! For instance, this is how it's done in the [geotoolbox](https://github.com/neocarto/geotoolbox/blob/e2cfa72f4a7ff2604397aa8d9dd5f815fb2443c6/src/gis/buffer.js#L48C33-L48C33). Especially if there are multiple instances of geos this should be preferable so the correct instance is chosen!

Okay, sure.
I will use the same way as geotoolbox. 👍

sanak commented:
Well, I tried 1st way in the following commit,
commit diff: https://github.com/chrispahm/geos-wasm/pull/19/commits/23dca351ef65b8767d483583d817aba4627592c4

but updating `README.md` - FAQ section may be a bit difficult (to prepare helpers method document outside of `geos` module)...

I will try 2nd way next week, because I will be a bit busy with other works until this weekend...
https://github.com/chrispahm/geos-wasm/pull/19#issuecomment-1763875189

sanak commented:
@chrispahm 
> Well, I tried 1st way in the following commit, commit diff: [23dca35](https://github.com/chrispahm/geos-wasm/commit/23dca351ef65b8767d483583d817aba4627592c4)
> 
> but updating `README.md` - FAQ section may be a bit difficult (to prepare helpers method document outside of `geos` module)...
> 
> I will try 2nd way next week, because I will be a bit busy with other works until this weekend... [#19 (comment)](https://github.com/chrispahm/geos-wasm/pull/19#issuecomment-1763875189)

I tried 1st way more before trying 2nd way, and I think that now it is okay, so could you review the latest code ?

Thanks,

chrispahm commented:
Hey @sanak, thanks for the update.

I've been reviewing the code and it looks good! 
Unfortunately, there is a last issue left with regards to bundling the code:

Right now, the helper modules are imported like this
```js
import { geojsonToGeosGeom } from '../../src/helpers/geojsonToGeosGeom.mjs'
import { geosGeomToGeojson } from '../../src/helpers/geosGeomToGeojson.mjs'
```

The `src` directory is stripped from the published npm package though (at least with our current settings). So, this would only work for people that installed the package from npm, but not for CDN users (in the browser). There is an [option in Rollup](https://rollupjs.org/configuration-options/#output-preservemodules) to preserve the original src structure, but I was wondering if it wouldn't be easier for users to just import the optional helpers like this
```js
import { geojsonToGeosGeom, geosGeomToGeojson } from 'geos-wasm/helpers'
```
at least that would be in line with what we discussed in #5, and also I think it's easier to remember!

If you approve of this, I can commit the necessary changes to your PR. Just wanted to check with you if that's ok

sanak commented:
@chrispahm Thanks for the comment!
> :
> Unfortunately, there is a last issue left with regards to bundling the code:
> 
> Right now, the helper modules are imported like this
> 
> ```js
> import { geojsonToGeosGeom } from '../../src/helpers/geojsonToGeosGeom.mjs'
> import { geosGeomToGeojson } from '../../src/helpers/geosGeomToGeojson.mjs'
> ```
> 
> The `src` directory is stripped from the published npm package though (at least with our current settings). So, this would only work for people that installed the package from npm, but not for CDN users (in the browser).

Oh, okay.
(I have no experience with CDN publishing yet, so I couldn't notice about that... 💦)

> There is an [option in Rollup](https://rollupjs.org/configuration-options/#output-preservemodules) to preserve the original src structure, but I was wondering if it wouldn't be easier for users to just import the optional helpers like this
> 
> ```js
> import { geojsonToGeosGeom, geosGeomToGeojson } from 'geos-wasm/helpers'
> ```
> 
> at least that would be in line with what we discussed in #5, and also I think it's easier to remember!

Oh, sure. I missed how bundle process works,
and yes, it is memorable. 😊

> If you approve of this, I can commit the necessary changes to your PR. Just wanted to check with you if that's ok

Yes, I approve the idea! 👍
And thanks for taking over about this! 🙇🙏

chrispahm commented:
Ok it's unfortunately not so easy to get submodules to work with ES6 imports... I'll ask around to see if there's a good solution

chrispahm commented:
Alright, finally got it to work through the use of [package.json sub-path exports](https://nodejs.org/api/packages.html#subpath-exports) 🎉 Please check again if this also works for you, then we can merge!

sanak commented:
@chrispahm Okay, thanks a lot! 👍🙇
The commit diff seems to be fine for me, and I confirmed that all test passed. 😄

In `README.md`, I noticed that my previous commit [helper methods link titles](https://github.com/sanak/geos-wasm/blob/feature/coordseq-copy-from-to-buffer/docs/index.md#what-is-the-preferred-way-to-pass-geometries-between-javascript-and-geos-wasm) are both `geojsonToGeosGeom`, but latter one needs to be modified `geosGeomToGeojson` as the link points, and that was the only problem... 🙇💦
Are you okay for me to add the fix commit ?

chrispahm reviewed (approved):
LGTM!

sanak commented:
@chrispahm 
> Hey @sanak just committed the change, sorry I misread your post and thought you asked if I could quickly do it! I'll merge this PR then if you don't have any further objections?

Thanks a lot! 🙇
I don't have any further objections. 👍
